DigitalLearningSolutions.Data/Services/LoginService.cs

```diff
@@ -1,65 +1,178 @@
+
+            if (!EmailIsTheSameOnAllAccounts(verifiedAdminUser, verifiedDelegateUsers))
+            {
+                throw new Exception("Not all accounts have the same email");
+            }
+
+            var adminAccountVerificationAttemptedAndFailed = unverifiedAdminUsers.Any() && verifiedAdminUser == null;
+            var doWeKnowWhichAdminAccountToRecordFailureAgainst = unverifiedAdminUsers.Count == 1;
+
+            var adminAccountIsAlreadyLocked = doWeKnowWhichAdminAccountToRecordFailureAgainst &&
+                                              unverifiedAdminUsers.Single().IsLocked;
+            var adminAccountHasJustBecomeLocked = doWeKnowWhichAdminAccountToRecordFailureAgainst &&
+                                                  unverifiedAdminUsers.Single().FailedLoginCount == 4 &&
+                                                  adminAccountVerificationAttemptedAndFailed;
+
+            var adminAccountIsLocked = adminAccountIsAlreadyLocked || adminAccountHasJustBecomeLocked;
+            var delegateAccountVerificationSuccessful = verifiedDelegateUsers.Any();
+            var shouldIncreaseFailedLoginCount =
+                doWeKnowWhichAdminAccountToRecordFailureAgainst && adminAccountVerificationAttemptedAndFailed &&
+                !delegateAccountVerificationSuccessful;
+
+            if (shouldIncreaseFailedLoginCount)
+            {
+                var adminUser = unverifiedAdminUsers.Single();
+                userService.IncrementFailedLoginCount(adminUser);
+            }
+
+            if (adminAccountIsLocked)
+            {
+                if (delegateAccountVerificationSuccessful)
+                {
+                    verifiedAdminUser = null;
+                }
+                else
+                {
+                    return new LoginResult(LoginAttemptResult.AccountLocked, unverifiedAdminUsers.Single());
+                }
+            }
+
+            if (verifiedAdminUser == null && !delegateAccountVerificationSuccessful)
+            {
+                return new LoginResult(LoginAttemptResult.InvalidPassword);
+            }
+
+            if (verifiedAdminUser != null)
+            {
+                userService.ResetFailedLoginCount(verifiedAdminUser);
+            }
+
+            var approvedVerifiedDelegates = verifiedDelegateUsers.Where(du => du.Approved).ToList();
+            if (verifiedAdminUser == null && !approvedVerifiedDelegates.Any())
+            {
+                return new LoginResult(LoginAttemptResult.AccountNotApproved);
+            }
+
+            var (verifiedLinkedAdmin, verifiedLinkedDelegates) = GetVerifiedLinkedAccounts(
+                password,
+                approvedVerifiedDelegates,
+                verifiedAdminUser
+            );
+
+            var adminUserToLoginIfCentreActive = new[] { verifiedAdminUser, verifiedLinkedAdmin }
+                .Where(au => au != null).Distinct().SingleOrDefault();

```

Comment:
They should either be the same or null, but this handles the case they are not. 
If we have 2 identical admins, but do not call Distinct before SingleOrDefault, it will cause an exception to occur.

Reply:
Just going to trust the code then, and remove the check. Since we are taking verifiedAdminUser ?? verifiedLinkedAdmin in the get linked accounts method, we will no only have a single AdminUser? to deal with.